File: problems/0211-design-add-and-search-words-data-structure/go/solution.go
package solution

// Design Add and Search Words — https://leetcode.com/problems/design-add-and-search-words-data-structure/
type WordDictionary struct {
	children map[byte]*WordDictionary
	end      bool
}

func Constructor() WordDictionary {
	return WordDictionary{children: make(map[byte]*WordDictionary)}
}

func (d *WordDictionary) AddWord(word string) {
	node := d
	for i := 0; i < len(word); i++ {
		c := word[i]
		if node.children[c] == nil {
			node.children[c] = &WordDictionary{children: make(map[byte]*WordDictionary)}
		}
		node = node.children[c]
	}
	node.end = true
}

func (d *WordDictionary) Search(word string) bool {
	var dfs func(node *WordDictionary, i int) bool
	dfs = func(node *WordDictionary, i int) bool {
		if i == len(word) {
			return node.end
		}
		c := word[i]
		if c == '.' {
			for _, child := range node.children {
				if dfs(child, i+1) {
					return true
				}
			}
			return false
		}
		next := node.children[c]
		if next == nil {
			return false
		}
		return dfs(next, i+1)
	}
	return dfs(d, 0)
}

func wordDictionaryOps(ops []string, args [][]string) []*bool {
	var wd WordDictionary
	out := []*bool{}
	for i, op := range ops {
		switch op {
		case "WordDictionary":
			wd = Constructor()
			out = append(out, nil)
		case "addWord":
			wd.AddWord(args[i][0])
			out = append(out, nil)
		case "search":
			v := wd.Search(args[i][0])
			out = append(out, &v)
		}
	}
	return out
}

File: problems/0211-design-add-and-search-words-data-structure/typescript/solution.ts
// Design Add and Search Words — https://leetcode.com/problems/design-add-and-search-words-data-structure/
class WordNode {
  children = new Map<string, WordNode>();
  end = false;
}

class WordDictionary {
  private root = new WordNode();

  addWord(word: string): void {
    let node = this.root;
    for (const c of word) {
      let next = node.children.get(c);
      if (next === undefined) {
        next = new WordNode();
        node.children.set(c, next);
      }
      node = next;
    }
    node.end = true;
  }

  search(word: string): boolean {
    const dfs = (node: WordNode, i: number): boolean => {
      if (i === word.length) return node.end;
      const c = word[i];
      if (c === ".") {
        for (const child of node.children.values()) {
          if (dfs(child, i + 1)) return true;
        }
        return false;
      }
      const next = node.children.get(c);
      if (next === undefined) return false;
      return dfs(next, i + 1);
    };
    return dfs(this.root, 0);
  }
}

export function wordDictionaryOps(ops: string[], args: string[][]): (boolean | null)[] {
  let wd = new WordDictionary();
  const out: (boolean | null)[] = [];
  for (let i = 0; i < ops.length; i++) {
    const arg = args[i];
    switch (ops[i]) {
      case "WordDictionary":
        wd = new WordDictionary();
        out.push(null);
        break;
      case "addWord":
        wd.addWord(arg[0]);
        out.push(null);
        break;
      case "search":
        out.push(wd.search(arg[0]));
        break;
    }
  }
  return out;
}
